### ISSUE TITLE
Layers could be slimmer by removing duplicated libraries

### ISSUE THREAD

**Issue by mnapoli** [mnapoli/bref#295]
It seems to me that many library files are duplicated in the layers. For example the folder `bref/lib` contains:

```
❯ ll lib
total 34080
drwxr-xr-x   4 matthieu  staff   128B  2 avr 20:30 cmake
-rwxr-xr-x   1 matthieu  staff   976B  2 avr 20:29 libcurl.la
-rwxr-xr-x   1 matthieu  staff   559K  2 avr 20:29 libcurl.so
-rwxr-xr-x   1 matthieu  staff   559K  2 avr 20:29 libcurl.so.4
-rwxr-xr-x   1 matthieu  staff   559K  2 avr 20:29 libcurl.so.4.5.0
-rw-r--r--   1 matthieu  staff   326K  2 avr 20:49 libpq.a
-rwxr-xr-x   1 matthieu  staff   246K  2 avr 20:31 libpq.so
-rwxr-xr-x   1 matthieu  staff   246K  2 avr 20:31 libpq.so.5
-rwxr-xr-x   1 matthieu  staff   246K  2 avr 20:31 libpq.so.5.9
-rw-r--r--   1 matthieu  staff   556K  2 avr 20:49 libsodium.a
-rwxr-xr-x   1 matthieu  staff   943B  2 avr 20:31 libsodium.la
-rwxr-xr-x   1 matthieu  staff   2,3M  2 avr 20:31 libsodium.so
-rwxr-xr-x   1 matthieu  staff   2,3M  2 avr 20:31 libsodium.so.23
-rwxr-xr-x   1 matthieu  staff   2,3M  2 avr 20:31 libsodium.so.23.1.0
-rwxr-xr-x   1 matthieu  staff   938B  2 avr 20:30 libxml2.la
-rwxr-xr-x   1 matthieu  staff   2,0M  2 avr 20:30 libxml2.so
-rwxr-xr-x   1 matthieu  staff   2,0M  2 avr 20:30 libxml2.so.2
-rwxr-xr-x   1 matthieu  staff   2,0M  2 avr 20:30 libxml2.so.2.9.8
-rwxr-xr-x   1 matthieu  staff   125K  2 avr 20:24 libz.so
-rwxr-xr-x   1 matthieu  staff   125K  2 avr 20:24 libz.so.1
-rwxr-xr-x   1 matthieu  staff   125K  2 avr 20:24 libz.so.1.2.11
drwxr-xr-x  22 matthieu  staff   704B  2 avr 20:49 php
drwxr-xr-x   7 matthieu  staff   224B  2 avr 20:31 pkgconfig
-rw-r--r--   1 matthieu  staff   226B  2 avr 20:30 xml2Conf.sh
```

For example I think those files:

```
libsodium.so
libsodium.so.23
```

should be symlinks to `libsodium.so.23.1.0` but they aren't (maybe #146? though not related to SAM here).

Because of that the layers are bigger than what they actually should be.

Could anyone else confirm that? Maybe @bubba-h57 

I was thinking of keeping only the `*.so` files and removing `*.so.*` files. WDYT?

**Comment by cebe** [comment]
Disclaimer: I have not looked into any more details of bref other than are provided in this issue so I might be wrong, but here is what I know :)

> I was thinking of keeping only the *.so files and removing *.so.* files.

depends on what your binaries (i.e. other programs using these libraries) link to. these version numbers are made to allow having more than one version of a lib installed. A binary can link to `libsodium.so` which means that whenever you update libsodium to a version that is not binary compatible with the version you had installed your program will be broken. for this reason you will probably have binaries linking to the concrete versions (`libsodium.so.23.1.0`) instead of the file without version number (`libsodium.so`).

So removing any of these is likely going to break your system (unless you want to check usage of these in every binary on the system, possible with tools like `ldd` and `objdump`). To reduce space you should convert them back into symlinks.

You can read more about this here:

- https://en.wikipedia.org/wiki/Soname
- https://unix.stackexchange.com/questions/475/how-do-so-shared-object-numbers-work

**Comment by bubba-h57** [comment]
Yes, as @cebe notes, the correct solution is using symlinks instead of copying the entire library. Bref originally decided to duplicate each library instead of using symlinks because of the desire to allow `sam invoke local` to work.

This was documented in https://github.com/mnapoli/bref/issues/146

So we can not simply delete the copies because the various libraries link to the different filenames, which denote versions. For example:

```bash
lrwxrwxrwx 1 sbx_user1051  495      16 Jan  8 17:15 libcurl.so -> libcurl.so.4.5.0
lrwxrwxrwx 1 sbx_user1051  495      16 Jan  8 17:15 libcurl.so.4 -> libcurl.so.4.5.0
-rwx------ 1 sbx_user1051  495  572239 Jan  8 17:15 libcurl.so.4.5.0
```

In this case, `libcurl.so.4.5.0` is the actualy library we built. It denotes the version as `major.minor.build == 4.5.0` so any consumer that requires linking to the `4.5` version of this library will link to this filename.

However, the majority of other consumers will simply want to link to a major build, and they will instead link to `libcurl.so.4`, which is normally a symlink to `libcurl.so.4.5.0` as in the example above.

Finally, some consumers don't care what version of the library they link to, just that they link to the library. These consumers will link to the `libcurl.so` filename, again a symlink to `libcurl.so.4.5.0`.

The best thing to do, if we desire slimmer layers, is to move back to symlinks in the package, but at this time that will break `sam local`.

Sorry I did not respond sooner, I missed the ping and had my head in the clouds, AWS Cloud to be precise. :-)
